docs/developer-docs/integrations/sns/get-sns/testflight.md

```diff
@@ -0,0 +1,258 @@
+in the canister code (the name is `MyGenericFunctions` and the description omitted in the sample code above).
+
+Once the proposal to register generic functions is accepted, you can submit proposals to execute them
+with a given binary payload:
+```
+sns-quill --canister-ids-file ./sns_canister_ids.json --pem-file $PEM_FILE make-proposal --proposal "(record { title=\"Execute generic functions.\"; url=\"https://example.com/\"; summary=\"This proposal executes generic functions.\"; action=opt variant {ExecuteGenericNervousSystemFunction = record {function_id=1000:nat64; payload=blob \"DIDL\01l\02\b9\fa\ee\18y\b5\f6\a1Cy\01\00\02\00\00\00\03\00\00\00\"}}})" $DEVELOPER_NEURON_ID > execute-generic-functions.json
+sns-quill send execute-generic-functions.json
+```
+The generic functions to be executed are specified by their numeric identifier defined in their registration proposal.
+The payload is a blob that is passed to both generic functions.
+The above sample payload `blob \"DIDL\01l\02\b9\fa\ee\18y\b5\f6\a1Cy\01\00\02\00\00\00\03\00\00\00\"` was obtained via
+```
+$ didc encode '(record {major=2:nat32; minor=3:nat32;})' --format blob
+blob "DIDL\01l\02\b9\fa\ee\18y\b5\f6\a1Cy\01\00\02\00\00\00\03\00\00\00"
+```
+and can be decoded as Candid payload (a record with two fields) in the canister Rust code
+```
+#[derive(CandidType, Debug, Deserialize)]
+struct Version {
+  major: u32,
+  minor: u32,
+}
+
+#[ic_cdk_macros::update]
+fn validate(x: Version) -> Result<String, String> {
+  // ...
+}
+```
+
+#### 8. Check the proposals of the testflight SNS.
+
+You can list all proposals in the testflight SNS as follows:
+```
+dfx canister call sns_governance list_proposals '(record {include_reward_status = vec {}; limit = 0; exclude_type = vec {}; include_status = vec {};})'
+```
+When running the testflight on mainnet, pass `--network ic` as an additional argument to `dfx canister`.
+You can also provide a limit and thus only obtain the last few proposals.
+
+#### 9. Aborting the testflight.
+
+As the developer keeps direct control over the registered dapp's canisters during the testflight,
+you can directly manage your dapp's canisters during the testflight if needed.
+However, we strongly encourage you to make sure you can also perform all required operations
+only via SNS proposals (possibly after upgrading your dapp).
+
+Once you are done with testing all kinds of SNS proposals needed to operate your dapp,
+you can finish the testflight.
+Make sure that you are a controller of all the canisters registered with the testflight SNS, e.g.,
+by invoking
+```
+dfx canister status test
+```
+When running the testflight on mainnet, pass `--network ic` as an additional argument to `dfx canister`.
+
+If this is the case, you can safely delete the SNS testflight canisters.
+Otherwise, you can restore control over your dapp's canisters by reinstalling
+the SNS root canister with the following code:
```

Comment:
Is this code somehow dependent on the tool versions and can go out of date?

Reply:
This recovery code essentially transfers control over a canister it controls to a specified principal. It is independent of tool versions.